cpp/src/copying/copy.cu

```diff
@@ -220,18 +229,31 @@ std::unique_ptr<column> scatter_gather_based_if_else(Left const& lhs,
     return std::move(result->release()[0]);
   }
 
-  // Bail out for Scalars.
-  // For nested types types, scatter/gather based copy_if_else() is not currently supported
-  // if either `lhs` or `rhs` is a scalar, partially because:
-  //   1. Struct scalars are not yet available.
-  //   2. List scalars do not yet support explosion to a full column.
-  CUDF_FAIL("Scalars of nested types are not currently supported!");
-  (void)lhs;
-  (void)rhs;
-  (void)size;
-  (void)is_left;
-  (void)stream;
-  (void)mr;
+  if constexpr (std::is_same<Left, cudf::scalar>::value &&
```

Comment:
Since there are now 4 paths generated here, in the interest of brevity, I wonder if we can reduce the code repetition, via a helper template.

I'll experiment, and offer a suggestion shortly.

Reply:
The other consideration is whether the "paths not taken" are actually eliminated by `nvcc`. Last I checked, both `nvcc-11` and `gcc-9` were good about this. But I'd feel better if we confirmed.

Reply:
The more I look at this, the more I think we should do this with iterators at some point, rather than inflate the struct scalar to a column. :/

Reply:
> Since there are now 4 paths generated here, in the interest of brevity, I wonder if we can reduce the code repetition, via a helper template.

Sorry for the delay. Here is the helper template I was suggesting:
```c++
template <typename T>
auto get_column_and_view(T const& t, 
                         cudf::size_type,
                         rmm::cuda_stream_view, 
                         rmm::mr::device_memory_resource*)
{
  return std::make_tuple(t, t);
}

template <>
auto get_column_and_view<cudf::scalar>(cudf::scalar const& scalar, 
                                   cudf::size_type size,
                                   rmm::cuda_stream_view stream, 
                                   rmm::mr::device_memory_resource* mr)
{
  auto column = cudf::make_column_from_scalar(scalar, size, stream, mr);
  return std::make_tuple(std::move(column), column->view());
}
```
This transparently returns a column and a view to it, regardless of whether the input is a `column_view` or a `scalar`.

`scatter_gather_based_if_else()` can now be reduced to:
```c++
template <typename Left, typename Right, typename Filter>
std::unique_ptr<column> scatter_gather_based_if_else(Left const& lhs,
                                                     Right const& rhs,
                                                     size_type size,
                                                     Filter is_left,
                                                     rmm::cuda_stream_view stream,
                                                     rmm::mr::device_memory_resource* mr)
{
  auto const [lhs_col, lhs_view] = get_column_and_view(lhs, size, stream, mr);
  auto const [rhs_col, rhs_view] = get_column_and_view(rhs, size, stream, mr);

  return scatter_gather_column_based_if_else(lhs_view, rhs_view, size, is_left, stream, mr);
}
```
`scatter_gather_column_based_if_else()` simply has the `column_view` based scatter-based `if-else`, without any `if constexpr` for anything.

Does this look agreeable?

Reply:
Please forgive my pointing out that `get_column_and_view()` has a couple of gotchas:

First, the function can't just return the column view. It must return the column, to keep the view alive.

Secondly, we can't `return std::make_tuple(std::move(column), column->view())` in one go. It's not reliable, because we don't want `column->view()` called *after* `std::move(column)`.
Switching the expressions around doesn't guarantee anything. The order of evaluation of those sub-expressions isn't guaranteed.

Reply:
> 
> 
> The more I look at this, the more I think we should do this with iterators at some point, rather than inflate the struct scalar to a column. :/

Yah, Gera and I talked about this offline, but the fundamental issue is that gather itself doesn't take iterators as column inputs (it does for the map though).  I think that's a pretty hard change, but it would be pretty nice. Doing so would also allow things like "multi-gather" where you can stitch together multiple columns together at once.